# Project - Final Report

## 1. Title

Project Title:
<br>
Bridging the Black Box: Explainable Anomaly Detection for Critical Infrastructure
<br> <br>
Course/Instructor Info:
<br>
Dr. Erdogan Dogdu
<br>
2025 Fall Semester - Introduction to Data Science (CS-4330-D10)
<br>
edogdu@angelo.edu
<br> <br>
Team info:
<br>
Serena Reese, Amiran Fields, Tatum Beaugh, Bill Mitchell and John DeLeon
<br> <br>
GitHub:
<br>
https://github.com/edogdu/CIS/tree/main

## 2. Project Description

Cyberattacks have increased in both number and complexity nationwide, and as critical infrastructure systems become more computerized, deep learning models have proven useful for detecting anomalies albeit their relatively uninterpretable nature. Our critical infrastructure framework seeks to incorporate explainable AI and knowledge graphs to improve anomaly detection, chiefly the context provided to the system’s users. Our framework includes an ETL process that takes network and sensor logs as input via a scalable pipeline powered by graph databases. An autoencoder then detects anomalies, while an XAI framework additionally provides context in terms of relevant nodes, features and relationships. An XGBoost model concludes by mapping the detected anomaly to cybersecurity standards to provide output in the form of recommended actions for the user to take.

## 3. Dataset

Name of the Dataset:
<br>
A hardware-in-the-loop water distribution testbed (WDT) dataset for cyber-physical security testing
<br> <br>
Date Published:
<br>
2010
<br> <br>
Last Updated:
<br>
Thu, 03/02/2023 - 17:22
<br> <br>
Data Source:
<br>
https://ieee-dataport.org/open-access/hardware-loop-water-distribution-testbed-wdt-dataset-cyber-physical-security-testing
<br> <br>
Data Format:
<br>
*.csv

### 3.1 - Overview of the Dataset

Our datasets were collected from an automated industrial monitoring system, specifically a cyber-physical Critical Infrastructure System (CIS) monitoring a water distribution testbed. Each physical data file contains time-stamped readings of tank levels, pump states, valve activations, flow sensor outputs and labeled system states. Each network data file contains time-stamped readings of MAC/IP pairs showing device-level and network-level behavior with port numbers for the CIS. Both will be used to detect volumetric and network anomalies to target data exfiltration attacks.

### 3.2 - Parts of the Dataset

testbed_system_1
> Network/csv <br>
  >> attack_1.csv (847.1 MB) <br>
  >> attack_2.csv (782.9 MB) <br>
  >> attack_3.csv (807.6 MB) <br>
  >> attack_4.csv (801.7 MB) <br>
  >> normal.csv (474 MB) <br>
  >> normal_1.csv (712 MB) <br>

> Physical <br>
  >> utf16 <br>
     >>> phy_att_1.csv (1.07 MB) <br>
     >>> phy_att_2.csv (953 KB) <br>
     >>> phy_att_3.csv (564 KB) <br>
     >>> phy_norm.csv (1.51 MB) <br>

  >> phy_att_1.csv (544 KB) <br>
  >> phy_att_2.csv (475 KB) <br>
  >> phy_att_3.csv (281 KB) <br>
  >> phy_att_4.csv (387 KB) <br>
  >> phy_norm.csv (770 KB) <br>

> CITATIONS.md <br>
> README.xlsx <br>

**Foglio 1 provided by the dataset:**
<br> <br>

| File             | Description                                          | Starting Time       | Ending Time         |
|------------------|------------------------------------------------------|---------------------|---------------------|
| normal.csv       | Acquisition of normal network traffic                | 09/04/2021 11:30:50 | 09/04/2021 11:27:57 |
| dataset_norm.csv | Acquisition of normal physical measures              | 09/04/2021 11:30:50 | 09/04/2021 12:27:57 |
| attack_1         | First acquisition of network traffic with attacks    | 09/04/2021 18:23:28 | 09/04/2021 19:03:47 |
| dataset_att_1    | First acquisition of physical measures under attack  | 09/04/2021 18:23:28 | 09/04/2021 19:03:47 |
| attack_2         | Second acquisition of network traffic with attacks   | 19/04/2021 15:37:12 | 19/04/2021 16:12:16 |
| dataset_att_2    | Second acquisition of physical measures under attack | 19/04/2021 15:37:12 | 19/04/2021 16:12:16 |
| attack_3         | Third acquisition of network traffic with attacks    | 09/04/2021 19:42:12 | 09/04/2021 20:03:05 |
| dataset_att_3    | Third acquisition of physical measures under attack  | 09/04/2021 19:42:12 | 09/04/2021 20:03:05 |
| attack_4         | Fourth acquisition of network traffic with attacks   | 21/02/2022 14:45:25 | 21/02/2022 15:13:55 |
| dataset_att_4    | Fourth acquisition of physical measures under attack | 21/02/2022 14:45:25 | 21/02/2022 15:13:55 |

### 3.3 - Features and Attributes

The WDT dataset contains 12 distinct cyberattack scenarios overall for intrusion detection research. It's provided in CSV format with numeric encoding of sensor/actuator values.

**Each physical file has the following features:**
<br> <br>
| Column Name                     | Description                                        | Type           | Example             | Source                                                                |
|---------------------------------|----------------------------------------------------|----------------|---------------------|-----------------------------------------------------------------------|
| Time                            | Timestamp                                          | datetime64[ns] | 09/04/2021 18:23:28 | Every sampling event records sensor and actuator states               |
| Tank_1 ... Tank_8               | Normalized water levels in tanks                   | int64          | 0                   | Water level sensors inside each storage tank                          |
| Pump_1 ... Pump_6               | Pump is running (True) or pump is off (False)      | bool           | False               | PLC boolean tag for pump state                                        |
| Flow_sensor_1 ... Flow_Sensor_4 | Raw pulses from the sensor to measure flow rate    | int64          | 0.233363            | Flow meter installed on a pipe                                        |
| Valv_1 ... Valv_22              | Valve closed (False) or valve open (True)          | bool           | False               | Actuator status bits from a PLC                                       |
| Label_n                         | Numeric-encoded label (0 = normal, 1 = attack)     | int64          | 0                   | Human-annotated/testbed-injected scenario logs                        |
| Label                           | Human-readable label (normal or anomaly)           | object         | normal              | Experiment metadata and offline labeling after cyber-attack injection |
| Label_encoded                   | Machine-learning-ready encoding of the label field | int64          | 0.098956            | Testbed scenario scripts/automatic labeling                           |

**Each network file has the following features:**
<br> <br>
| Column Name     | Description                                       | Type    | Example                    | Source                                               |
|-----------------|---------------------------------------------------|---------|----------------------------|------------------------------------------------------|
| Time            | Timestamp of the packet/flow                      | object  | 2021-04-09 18:23:28.385003 | Packet capture (PCAP) system                         |
| mac_s           | Source MAC address (hardware identifier)          | object  | 74:46:a0:bd:a7:1b          | Network interface card (NIC) of source device        |
| mac_d           | Destination MAC address                           | object  | 0a:fe:ec:47:74:fb          | NIC of receiving device                              |
| ip_s            | Source IP address                                 | object  | 84.3.251.20                | Address of packet-sending device                     |
| ip_d            | Destination IP address                            | object  | 84.3.251.102               | Address of packet-receiving device                   |
| sport           | Source port number                                | float64 | 56667.0                    | Ephemeral TCP port                                   |
| dport           | Destination port number                           | float64 | 502.0                      | Modbus/TCP service port on a PLC                     |
| proto           | Protocol used, like Modbus or TCP                 | object  | Modbus                     | Application-layer industrial protocol                |
| flags           | Protocol-specific flags                           | float64 | 11000.0                    | TCP flags extracted from the network packet          |
| size            | Size of the packet/message in bytes               | int64   | 66                         | Packet size from packet capture                      |
| modbus_fn       | Modbus function code or operation, like Request   | object  | Read Coils Request         | Modbus function code parsed from Modbus/TCP payload  |
| n_pkt_src       | Number of packets sent from source to destination | float64 | 0.0                        | Count of packets previously sent by this device      |
| n_pkt_dst       | Number of packets sent from destination to source | float64 | 0.0                        | Count of packets previously sent to this destination |
| modbus_response | Response payload from a Modbus device             | object  | [0]                        | Response payload from PLC                            |
| label_n         | Numeric label (0 or 1)                            | int64   | 0                          | Numeric label assigned by dataset authors            |
| label           | Human-readable label                              | object  | normal                     | Human-readable label for anomaly classification      |

### 3.4 - Size of the Dataset

In [ ]:
import pandas as pd
import gc

In [ ]:
# Declare paths
network_path = "/content/drive/MyDrive/CS/AI Cybersecurity/CIS_datasets/NetworkDataSet/"
physical_path = "/content/drive/MyDrive/CS/AI Cybersecurity/CIS_datasets/PhysicalDataSet/"

The dataset is provided via dataset.zip, which is 859.78 MB in size. The dataset provides the 12 distinct attack scenarios by distributing them across the four network attack files that correspond to the physical measurements in the four physical attack files. In other words, the physical CSVs are smaller subsets that correspond to the same attack scenarios, but instead capture the physical process variables associated with the system atttacks rather than the network traffic. The normal files contain only baseline system operation without any attacks. Each record in all attack files are tagged with whether it belongs to a normal or attack scenario, so the data can be filtered by scenario ID once loaded into an IDE.

| File          | Content               | Number of Features | Attack Subsets | Rows    | Columns | Delimiter | Encoding |
|---------------|-----------------------|--------------------|----------------|---------|---------|-----------|----------|
| attack_1.csv  | Network logs          | 16                 | 8              | 5527409 | 16      | ,         | utf-8    |
| attack_2.csv  | Network logs          | 16                 | 13             | 5159469 | 16      | ,         | utf-8    |
| attack_3.csv  | Network logs          | 16                 | 7              | 5862547 | 16      | ,         | utf-8    |
| attack_4.csv  | Network logs          | 16                 | 14             | 5522490 | 16      | ,         | utf-8    |
| normal.csv    | Network logs          | 16                 | 0              | 7757289 | 16      | ,         | utf-8    |
| phy_att_1.csv | Physical process logs | 43                 | 8              | 2420    | 43      | \t        | utf-16   |
| phy_att_2.csv | Physical process logs | 43                 | 13             | 2104    | 43      | \t        | utf-16   |
| phy_att_3.csv | Physical process logs | 43                 | 7              | 1254    | 43      | \t        | utf-16   |
| phy_att_4.csv | Physical process logs | 43                 | 14             | 1717    | 43      | ,         | utf-8    |
| phy_norm.csv  | Physical process logs | 43                 | 0              | 3428    | 43      | \t        | utf-16   |

In [ ]:
gc.collect()

0

In [ ]:
# Define file list
files = [
    "attack_1.csv", "attack_2.csv", "attack_3.csv", "attack_4.csv",
    "normal.csv",
    "phy_att_1.csv", "phy_att_2.csv", "phy_att_3.csv", "phy_att_4.csv",
    "phy_norm.csv"
]

In [ ]:
# Collect metadata
meta = []
network_counts = []
physical_counts = []

for f in files:
    try:
        # Use utf-16 for physical files, utf-8 otherwise
        if f == "phy_att_4.csv":
          df = pd.read_csv(physical_path + f, sep=",", encoding="utf-8")
          rows, cols = df.shape
          physical_counts.append(rows)
        else:
          if f.startswith("phy_"):
              df = pd.read_csv(physical_path + f, sep="\t", encoding="utf-16")
              rows, cols = df.shape
              physical_counts.append(rows)
          else:
              df = pd.read_csv(network_path + f)
              rows, cols = df.shape
              network_counts.append(rows)

        meta.append({"file": f, "rows": rows, "cols": cols})
        print(f"\n{f} --> Columns:")
        print(list(df.columns))

    except Exception as e:
        print(f"Error reading {f}: {e}")


attack_1.csv --> Columns:
['Time', ' mac_s', ' mac_d', ' ip_s', ' ip_d', ' sport', ' dport', ' proto', ' flags', ' size', ' modbus_fn', ' n_pkt_src', ' n_pkt_dst', ' modbus_response', ' label_n', ' label']

attack_2.csv --> Columns:
['Time', ' mac_s', ' mac_d', ' ip_s', ' ip_d', ' sport', ' dport', ' proto', ' flags', ' size', ' modbus_fn', ' n_pkt_src', ' n_pkt_dst', ' modbus_response', ' label_n', ' label']

attack_3.csv --> Columns:
['Time', ' mac_s', ' mac_d', ' ip_s', ' ip_d', ' sport', ' dport', ' proto', ' flags', ' size', ' modbus_fn', ' n_pkt_src', ' n_pkt_dst', ' modbus_response', ' label_n', ' label']

attack_4.csv --> Columns:
['Time', 'mac_s', 'mac_d', 'ip_s', 'ip_d', 'sport', 'dport', 'proto', 'flags', 'size', 'modbus_fn', 'modbus_response', 'n_pkt_src', 'n_pkt_dst', 'label_n', 'label']

normal.csv --> Columns:
['Time', 'mac_s', 'mac_d', 'ip_s', 'ip_d', 'sport', 'dport', 'proto', 'flags', 'size', 'modbus_fn', 'n_pkt_src', 'n_pkt_dst', 'modbus_response', 'label_n', 'label

In [ ]:
# Compute averages
avg_network = sum(network_counts) / len(network_counts)
avg_physical = sum(physical_counts) / len(physical_counts)

print("\nAverage rows (network files):", avg_network)
print("Average rows (physical files):", avg_physical)


Average rows (network files): 5965840.8
Average rows (physical files): 2184.6


In [ ]:
# Compute totals and averages
meta_df = pd.DataFrame(meta)

# Total rows across all files
total_rows = meta_df['rows'].sum()

# Average rows per file
avg_rows = meta_df['rows'].mean()

# Number of features per log (columns per file)
features_per_file = meta_df[['file','cols']]

print("Total rows:", total_rows)
print("Average rows per file:", avg_rows)
print("Features per file:\n", features_per_file)

Total rows: 29840127
Average rows per file: 2984012.7
Features per file:
             file  cols
0   attack_1.csv    16
1   attack_2.csv    16
2   attack_3.csv    16
3   attack_4.csv    16
4     normal.csv    16
5  phy_att_1.csv    43
6  phy_att_2.csv    43
7  phy_att_3.csv    43
8  phy_att_4.csv    43
9   phy_norm.csv    43


Average number of rows for a network file:
<br>
5965840.8
<br> <br>
Average number of rows for a physical file:
<br>
2184.6
<br> <br>
Total rows across all files:
<br>
29840127
<br> <br>
Number of features per network file:
<br>
16
<br> <br>
Number of features per physical file:
<br>
43

In [ ]:
gc.collect()

0

### 3.5 - Data Quality Issues

Load one network dataframe and one physical dataframe and check for missing values

In [ ]:
# Load one network file, utf-8 and ,
network_df = pd.read_csv(network_path + "normal.csv")
network_df.columns = network_df.columns.str.strip().str.lower()   # standardize cols

# Check shape
print("Network file rows and columns:")
print("Rows:", network_df.shape[0], "Cols:", network_df.shape[1])
missing_percent = network_df.isnull().mean() * 100
print(missing_percent)

Network file rows and columns:
Rows: 7757289 Cols: 16
time                0.000000
mac_s               0.000000
mac_d               0.000000
ip_s                0.000000
ip_d                0.000000
sport               0.000000
dport               0.000000
proto               0.000000
flags               0.000000
size                0.000000
modbus_fn           2.710264
n_pkt_src           0.000000
n_pkt_dst           0.000000
modbus_response    51.353508
label_n             0.000000
label               0.000000
dtype: float64


In [ ]:
# Check for duplicates, unexpected values/encoding issues, outliers and whitespace/typos in column names
print("Duplicate rows:", network_df.duplicated().sum()) # duplicates
print(network_df.describe())  # summary statistics

Duplicate rows: 0
              sport         dport         flags          size     n_pkt_src  \
count  7.757289e+06  7.757289e+06  7.757289e+06  7.757289e+06  7.757289e+06   
mean   3.084898e+04  3.072654e+04  1.091852e+04  6.532566e+01  3.069325e+01   
std    3.034108e+04  3.036678e+04  8.315406e+02  1.389839e+00  1.752479e+01   
min    2.500000e+01  2.500000e+01  1.000000e+01  6.000000e+01  0.000000e+00   
25%    5.020000e+02  5.020000e+02  1.100000e+04  6.500000e+01  1.500000e+01   
50%    3.503500e+04  5.020000e+02  1.100000e+04  6.600000e+01  1.900000e+01   
75%    6.151600e+04  6.151600e+04  1.100000e+04  6.600000e+01  5.000000e+01   
max    6.151700e+04  6.151700e+04  1.100000e+04  7.800000e+01  5.100000e+01   

          n_pkt_dst    label_n  
count  7.757289e+06  7757289.0  
mean   3.073045e+01        0.0  
std    1.749213e+01        0.0  
min    0.000000e+00        0.0  
25%    1.500000e+01        0.0  
50%    1.900000e+01        0.0  
75%    5.000000e+01        0.0  
max   

In [ ]:
3# Count missing values per column for network data
print(network_df.isnull().sum())

# Total missing values in the whole dataset
print("Total missing:", network_df.isnull().sum().sum())

time                     0
mac_s                    0
mac_d                    0
ip_s                     0
ip_d                     0
sport                    0
dport                    0
proto                    0
flags                    0
size                     0
modbus_fn           210243
n_pkt_src                0
n_pkt_dst                0
modbus_response    3983640
label_n                  0
label                    0
dtype: int64
Total missing: 4193883


In [ ]:
# Load one physical file, only phy_att_4 is , and utf-8
physical_df = pd.read_csv(physical_path + "phy_norm.csv", sep="\t", encoding="utf-16")
physical_df.columns = physical_df.columns.str.strip().str.lower()   # standardize cols

print("Physical file rows and columns:")
print("Rows:", physical_df.shape[0], "Cols:", physical_df.shape[1])
missing_percent = physical_df.isnull().mean() * 100
print(missing_percent)

Physical file rows and columns:
Rows: 3428 Cols: 43
time             0.0
tank_1           0.0
tank_2           0.0
tank_3           0.0
tank_4           0.0
tank_5           0.0
tank_6           0.0
tank_7           0.0
tank_8           0.0
pump_1           0.0
pump_2           0.0
pump_3           0.0
pump_4           0.0
pump_5           0.0
pump_6           0.0
flow_sensor_1    0.0
flow_sensor_2    0.0
flow_sensor_3    0.0
flow_sensor_4    0.0
valv_1           0.0
valv_2           0.0
valv_3           0.0
valv_4           0.0
valv_5           0.0
valv_6           0.0
valv_7           0.0
valv_8           0.0
valv_9           0.0
valv_10          0.0
valv_11          0.0
valv_12          0.0
valv_13          0.0
valv_14          0.0
valv_15          0.0
valv_16          0.0
valv_17          0.0
valv_18          0.0
valv_19          0.0
valv_20          0.0
valv_21          0.0
valv_22          0.0
label_n          0.0
label            0.0
dtype: float64


In [ ]:
# Count missing values per column for physical data
print(physical_df.isnull().sum())

# Total missing values in the whole dataset
print("Total missing:", physical_df.isnull().sum().sum())

time             0
tank_1           0
tank_2           0
tank_3           0
tank_4           0
tank_5           0
tank_6           0
tank_7           0
tank_8           0
pump_1           0
pump_2           0
pump_3           0
pump_4           0
pump_5           0
pump_6           0
flow_sensor_1    0
flow_sensor_2    0
flow_sensor_3    0
flow_sensor_4    0
valv_1           0
valv_2           0
valv_3           0
valv_4           0
valv_5           0
valv_6           0
valv_7           0
valv_8           0
valv_9           0
valv_10          0
valv_11          0
valv_12          0
valv_13          0
valv_14          0
valv_15          0
valv_16          0
valv_17          0
valv_18          0
valv_19          0
valv_20          0
valv_21          0
valv_22          0
label_n          0
label            0
dtype: int64
Total missing: 0


In [ ]:
# Check for duplicates, unexpected values/encoding issues, outliers and whitespace/typos in column names
print("Duplicate rows:", physical_df.duplicated().sum()) # duplicates
print(physical_df.describe())  # summary statistics

Duplicate rows: 0
            tank_1       tank_2       tank_3       tank_4       tank_5  \
count  3428.000000  3428.000000  3428.000000  3428.000000  3428.000000   
mean    710.314761   732.337223   597.789965   356.985123   368.955368   
std     663.274738   647.241518   467.890372   369.808391   271.544093   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       5.000000     5.000000     9.000000    31.750000    93.750000   
50%     621.500000   669.000000   621.000000   180.000000   384.000000   
75%    1284.250000  1292.250000  1018.000000   688.000000   604.250000   
max    1994.000000  1985.000000  1513.000000  1143.000000   831.000000   

            tank_6       tank_7       tank_8  flow_sensor_1  flow_sensor_2  \
count  3428.000000  3428.000000  3428.000000    3428.000000    3428.000000   
mean    212.233081   183.305426   190.320595    1605.600933    1652.275379   
std     166.565724   185.233377   184.099894    1961.012797    1969.827279   
min

| File          | Duplicate Rows | Missing Data Percentages                                                                                                                                                                                 | Missing Data Counts                                                                                                                                               | Total Missing Values | Missing Percentage of Dataset |
|---------------|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------|-------------------------------|
| attack_1.csv  | 2              | ip_s (0.008594%), ip_d (0.008594%), sport (0.009317%),<br>dport (0.009317%), flags (0.009317%), modbus_fn (2.770249%),<br>n_pkt_src (0.008594%), n_pkt_dst (0.008594%), modbus_response (51.383605%)     | ip_s (475), ip_d (475), sport (515),<br>dport (515), flags (515), modbus_fn (153123),<br>n_pkt_src (475), n_pkt_dst (475), modbus_response (2840182)              | 2996750              | ~3.39%                        |
| attack_2.csv  | 349            | ip_s (0.005349%), ip_d (0.005349%), sport (7.469431%),<br>dport (7.469431%), flags (7.469431%), modbus_fn (10.456832%),<br>n_pkt_src (0.005349%), n_pkt_dst (0.005349%), modbus_response (55.227195%     | ip_s (276), ip_d (276), sport (385383),<br>dport (385383), flags (385383), modbus_fn (539517),<br>n_pkt_src (276), n_pkt_dst (276), modbus_response (2849430)     | 4546200              | ~5.51%                        |
| attack_3.csv  | 118270         | ip_s (0.003224%), ip_d (0.003224%), sport (52.312365%),<br>dport (52.312365%), flags (52.312365%), modbus_fn (53.627544%),<br>n_pkt_src (0.003224%), n_pkt_dst (0.003224%), modbus_response (76.812501%) | ip_s (189), ip_d (189), sport (3066837),<br>dport (3066837), flags (3066837), modbus_fn (3143940),<br>n_pkt_src (189), n_pkt_dst (189), modbus_response (4503169) | 16848376             | ~17.95%                       |
| attack_4.csv  | 391381         | ip_s (0.007949%), ip_d (0.007949%), sport (21.227834%),<br>dport (21.227834%), flags (21.227834%), modbus_fn (31.722864%),<br>modbus_response (65.859259%), n_pkt_src (0.007949%), n_pkt_dst (0.007949%  | ip_s (439), ip_d (439), sport (1172305),<br>dport (1172305), flags (1172305), modbus_fn (1751892),<br>modbus_response (3637071), n_pkt_src (439), n_pkt_dst (439) | 8907634              | ~10.08%                       |
| normal.csv    | 0              | modbus_fn (2.710264%), modbus_response (51.353508)                                                                                                                                                       | modbus_fn (210243), modbus_response (3983640)                                                                                                                     | 4193883              | ~3.38%                        |
| phy_att_1.csv | 0              | 0%                                                                                                                                                                                                       | 0                                                                                                                                                                 | 0                    | 0%                            |
| phy_att_2.csv | 0              | 0%                                                                                                                                                                                                       | 0                                                                                                                                                                 | 0                    | 0%                            |
| phy_att_3.csv | 0              | 0%                                                                                                                                                                                                       | 0                                                                                                                                                                 | 0                    | 0%                            |
| phy_att_4.csv | 0              | 0%                                                                                                                                                                                                       | 0                                                                                                                                                                 | 0                    | 0%                            |
| phy_norm.csv  | 0              | 0%                                                                                                                                                                                                       | 0                                                                                                                                                                 | 0                    | 0%                            |

Missing Time Steps

The Time column has mixed formats, some rows including fractional seconds and some not. Here's how the timestamps are stored in our network logs. Instead of clean 1-second steps, the 'Time' column records events with microsecond precision (fractions of a second). Our network logs don't have a fixed interval, but rather are event-driven.

Our physical logs instead show that every consecutive timestamp is separated by 1 second with only one unique interval. Our network logs are packet-driven, and our physical logs are time-driven, sampled once per second.

In [ ]:
# Flag missing time steps across physical files
physical_files = [
    "phy_att_1.csv", "phy_att_2.csv", "phy_att_3.csv",
    "phy_att_4.csv", "phy_norm.csv"
]

for f in physical_files:
    try:
        # Load with UTF-16 encoding and tab delimiter for all except phy_att_4.csv
        if f == "phy_att_4.csv":
          df = pd.read_csv(physical_path + f, sep=",", encoding="utf-8")
          df['Time'] = pd.to_datetime(df['Time'], dayfirst=True, errors='coerce')
        else:
          df = pd.read_csv(physical_path + f, sep="\t", encoding="utf-16")
          df['Time'] = pd.to_datetime(df['Time'], dayfirst=True, errors='coerce')

        # Calculate differences between consecutive timestamps
        diffs = df['Time'].diff().dropna()

        # Count how many gaps are larger than 1 second
        missing_steps = (diffs > pd.Timedelta(seconds=1)).sum()

        print(f"\n{f}")
        print("Rows:", df.shape[0])
        print("Unique intervals:", diffs.unique())
        print("Missing steps:", missing_steps)

    except Exception as e:
        print(f"Error reading {f}: {e}")


phy_att_1.csv
Rows: 2420
Unique intervals: <TimedeltaArray>
['0 days 00:00:01']
Length: 1, dtype: timedelta64[ns]
Missing steps: 0

phy_att_2.csv
Rows: 2104
Unique intervals: <TimedeltaArray>
['0 days 00:00:01', '0 days 00:00:02']
Length: 2, dtype: timedelta64[ns]
Missing steps: 1

phy_att_3.csv
Rows: 1254
Unique intervals: <TimedeltaArray>
['0 days 00:00:01']
Length: 1, dtype: timedelta64[ns]
Missing steps: 0

phy_att_4.csv
Rows: 1717
Unique intervals: <TimedeltaArray>
['0 days 00:00:01']
Length: 1, dtype: timedelta64[ns]
Missing steps: 0

phy_norm.csv
Rows: 3428
Unique intervals: <TimedeltaArray>
['0 days 00:00:01']
Length: 1, dtype: timedelta64[ns]
Missing steps: 0


Results:
<br>
phy_att_1.csv --> perfectly aligned, no missing steps
<br>
phy_att_2.csv --> one gap of 2 seconds, 1 missing step
<br>
phy_att_3.csv --> perfectly aligned
<br>
phy_att_4.csv --> perfectly aligned
<br>
phy_norm.csv --> perfectly aligned

*Are there any large gaps in packet activity for network logs?*

In [ ]:
network_files = [
    "attack_1.csv", "attack_2.csv", "attack_3.csv",
    "attack_4.csv", "normal.csv"
]

for f in network_files:
    try:
        # Load network logs (UTF-8, comma-separated)
        df = pd.read_csv(network_path + f)
        df.columns = df.columns.str.strip().str.lower()

        # Convert time column
        df['time'] = pd.to_datetime(df['time'], errors='coerce')

        # Compute time differences
        diffs = df['time'].diff().dropna()

        # Define what counts as a "gap"
        # Here: anything > 1 second
        missing_steps = (diffs > pd.Timedelta(seconds=1)).sum()

        print(f"\n{f}")
        print("Rows:", df.shape[0])
        print("Unique intervals (sample):", diffs.unique()[:5])
        print("Total unique intervals:", len(diffs.unique()))
        print("Gaps > 1 second:", missing_steps)

    except Exception as e:
        print(f"Error reading {f}: {e}")


attack_1.csv
Rows: 5527409
Unique intervals (sample): <TimedeltaArray>
['0 days 00:00:00.000002', '0 days 00:00:00.000001', '0 days 00:00:00.000478',
 '0 days 00:00:00.000014', '0 days 00:00:00.000855']
Length: 5, dtype: timedelta64[ns]
Total unique intervals: 3399
Gaps > 1 second: 0

attack_2.csv
Rows: 5159469
Unique intervals (sample): <TimedeltaArray>
['0 days 00:00:00.001427', '0 days 00:00:00.000004', '0 days 00:00:00.000002',
 '0 days 00:00:00.000296', '0 days 00:00:00.000001']
Length: 5, dtype: timedelta64[ns]
Total unique intervals: 3390
Gaps > 1 second: 0

attack_3.csv
Rows: 5862547
Unique intervals (sample): <TimedeltaArray>
['0 days 00:00:00.002258', '0 days 00:00:00.000016', '0 days 00:00:00.000001',
        '0 days 00:00:00', '0 days 00:00:00.000294']
Length: 5, dtype: timedelta64[ns]
Total unique intervals: 3661
Gaps > 1 second: 0

attack_4.csv
Rows: 5522490
Unique intervals (sample): <TimedeltaArray>
['0 days 00:00:00.000003', '0 days 00:00:00.000028', '0 days 00:00:00.

*Network gaps can reveal attack effects, switch congestion, logging failures, packet loss and system resets.*
<br> <br>
Results:
<br>
attack_1 --> 5.5M rows, 3399 unique intervals, 0 gaps
<br>
attack_2 --> 5.1M rows, 3390 unique intervals, 0 gaps
<br>
attack_3 --> 5.8M rows, 3661 unique intervals, 0 gaps
<br>
attack_4 --> 5.5M rows, 3843 unique intervals, 0 gaps
<br>
normal --> 7.7M rows, 3370 unique intervals, 0 gaps

attack_1 shows dense packet activity with no outages. attack_2 shows the same pattern, indicating stable logging. attack_3 shows slightly more interval variety, but it's still continuous. attack_4 shows the highest variability, but there's still no gaps. normal shows the very stable baseline traffic of our system.

## 4. Data Cleanup

### 4.1 - Initial Data Cleaning

We only have one missing timestep in phy_att_2.csv. We have a considerable amount of missing data in network logs, but not in physical logs. We have inconsistent timestamp formatting between physical and network logs, along with some seemingly duplicate network records.

**Calculate Misaligned Timestamps**

Resample the network logs to 1-second bins to get packet counts per second. Use this to align network and physical logs perfectly.

Issue detected:
1. Our network timestamps use YYYY-MM-DD, while our physical timestamps use YYYY-DD-MM.

In [ ]:
# Force Pandas to interpret the physical timestamps as day-first
physical_df['time'] = pd.to_datetime(
    physical_df['time'],
    dayfirst=True,
    errors='coerce'
)

# Force Pandas to interpret normal.csv as NOT day-first
network_df['time'] = pd.to_datetime(
    network_df['time'],
    dayfirst=False,
    errors='coerce'
)

# Set the indexes
physical_df = physical_df.set_index('time').sort_index()
network_df = network_df.set_index('time').sort_index()

In [ ]:
# Verify the network file and physical file timestamps have the same format
print(network_df.index.min(), network_df.index.max())
print(physical_df.index.min(), physical_df.index.max())

2021-04-09 11:30:52.716203 2021-04-09 12:27:56.049946
2021-04-09 11:30:50 2021-04-09 12:27:57


In [ ]:
# Resample network logs to 1-second bins
network_1s = network_df.resample('1s').size().rename("packet_count")

*Now, we have acquired timestamp --> number of packets in that second.* Physical logs are sampled at 1s already, so just ensure the timestamp is parsed and set as the index.

Next, merge the two datasets on their newly shared 1-second timemstamps.
<br>
Guide:
1. inner keeps only timestamps present in both files
2. left keeps all physical timestamps
3. outer keeps everything (use for debugging)

In [ ]:
# Align the two datasets
aligned = physical_df.join(network_1s, how='inner')

In [ ]:
# Inspect the merged dataset
print(aligned.head())
print(aligned.shape)

                     tank_1  tank_2  tank_3  tank_4  tank_5  tank_6  tank_7  \
time                                                                          
2021-04-09 11:30:52       0       0       0       0       0       0       0   
2021-04-09 11:30:53       0       0       0       0       0       0       0   
2021-04-09 11:30:54       0       0       0       0       0       0       0   
2021-04-09 11:30:55       0       0       0       0       0       0       0   
2021-04-09 11:30:56       1       0       0       0       0       0       0   

                     tank_8  pump_1  pump_2  ...  valv_16  valv_17  valv_18  \
time                                         ...                              
2021-04-09 11:30:52       0   False   False  ...    False    False    False   
2021-04-09 11:30:53       0   False   False  ...    False    False    False   
2021-04-09 11:30:54       0   False   False  ...    False    False    False   
2021-04-09 11:30:55       0   False   False  ...   

**Now, we have an aligned physical + network dataset.**

Label consistency

In [ ]:
# Check first few rows around the start of an attack
aligned[['label', 'packet_count']].head(20)

,label,packet_count
time,,
2021-04-09 11:30:52,normal,633
2021-04-09 11:30:53,normal,2235
2021-04-09 11:30:54,normal,2176
2021-04-09 11:30:55,normal,2287
2021-04-09 11:30:56,normal,2274
2021-04-09 11:30:57,normal,2275
2021-04-09 11:30:58,normal,2278
2021-04-09 11:30:59,normal,2259
2021-04-09 11:31:00,normal,2288


In [ ]:
# Check for label drift
aligned['label_change'] = aligned['label'].ne(aligned['label'].shift())
print(aligned[aligned['label_change']].head())

                     tank_1  tank_2  tank_3  tank_4  tank_5  tank_6  tank_7  \
time                                                                          
2021-04-09 11:30:52       0       0       0       0       0       0       0   

                     tank_8  pump_1  pump_2  ...  valv_17  valv_18  valv_19  \
time                                         ...                              
2021-04-09 11:30:52       0   False   False  ...    False    False    False   

                     valv_20  valv_21  valv_22  label_n   label  packet_count  \
time                                                                            
2021-04-09 11:30:52    False    False    False        0  normal           633   

                     label_change  
time                               
2021-04-09 11:30:52          True  

[1 rows x 44 columns]


In [ ]:
# Check for missing packet counts after the join
aligned['packet_count'].isna().sum()

np.int64(0)

Labels are synchronized, and there are no missing packet counts.

In [ ]:
gc.collect()

9175

To clean the dataset, we:
<br> <br>
*Standardized and parsed timestamps*
1. Loaded timestamps from both the physical process logs and the network packet logs
2. Normalized formats, so both physical and network logs use the same structure, especially dates and column titles
3. Ensure Pandas interpreted dates correctly
4. Converted timestamps to a uniform timezone to align datasets
5. Set timestamps as the DataFrame index for time-series operations

<br>

*Cleaned column names*
1. Stripped whitespace, BOM characters and inconsistent capitalization
2. Converted all column names to lowercase for consistency

<br>

*Resampled network logs to 1-second resolution*
1. Network logs were event-driven, so we aggregated them into 1-second bins using .resample('1S').count()
2. Produced a clean packet_count column representing packets per second

<br>

*Merged physical and network logs*
1. Joined the physical dataset and the resampled network dataset on timestamp
2. Used an inner join to keep only timestamps present in both datasets
3. Verified that the merged dataset had no missing seconds

<br>

*Handled missing values*
1. Checked for missing packet counts after merging
2. Found zero missing values, so no filling was required
3. For specifically physical sensors, confirmed no missing values in the aligned window were present

<br>

*Removed impossible or invalid sensor values*
1. Identified and removed negative tank levels, negative flow readings and physically impossible spikes

<br>

*Deduplicated log entries*
1. Checked for duplicate timestamps or repeated rows
2. Dropped duplicates using 'keep=first'

<br>

*Converted categorical fields to boolean/integer*
1. Standardized into consistent numeric or boolean values

<br>

*Verified label synchronization*
1. Checked whether attack levels were sparse or per-second
2. Confirmed labels were already synchronous and continuous
3. Verified label transitions matched actual attack starting times

Our final aligned dataframe should include:
1. Aligned timestamps
2. Resampled network data points
3. Merged physical and network data
4. Consistent labels

In [ ]:
# Save the aligned dataset
aligned.to_csv("aligned_norm.csv")

### 4.2 - Feature Engineering

In [ ]:
# Load and combine all aligned files
import glob

# Path pattern for all aligned CSVs
files = glob.glob("aligned_*.csv")

dfs = []
for f in files:
    df = pd.read_csv(f, parse_dates=['time'], index_col='time')
    df['scenario'] = f  # optional: track which file each row came from
    dfs.append(df)

# Combine into one dataset
full_aligned_dataset = pd.concat(dfs).sort_index()

In [ ]:
# Verify correct loading
print(full_aligned_dataset.shape)
print(full_aligned_dataset['label'].value_counts())
print(full_aligned_dataset['scenario'].value_counts())

(10903, 60)
label
normal            8637
MITM              1008
physical fault     685
DoS                310
nomal              249
scan                14
Name: count, dtype: int64
scenario
aligned_norm.csv        3425
aligned_attack_1.csv    2420
aligned_attack_2.csv    2096
aligned_attack_4.csv    1710
aligned_attack_3.csv    1252
Name: count, dtype: int64


This new full aligned dataset for our CIS contains 10.903 rows across all scenarios plus 60 features (physical + network + engineered labels). We have multiple attack types, and majority label is normal. We have a typo however - "nomal".

In [ ]:
# Fix typo of "nomal"
full_aligned_dataset['label'] = full_aligned_dataset['label'].replace({'nomal': 'normal'})

In [ ]:
# Verify label consistency
full_aligned_dataset['label'].value_counts()

,count
label,
normal,8886
MITM,1008
physical fault,685
DoS,310
scan,14


**Scenario Distribution in Full Aligned Dataset**
| aligned_norm.csv     | 3425 |
|----------------------|------|
| aligned_attack_1.csv | 2420 |
| aligned_attack_2.csv | 2096 |
| aligned_attack_3.csv | 1252 |
| aligned_attack_4.csv | 1710 |

### 4.3 - Data Selection

To detect anomalies, we'll use:
1. All physical sensors (tank levels, flow sensors, pressure sensors, pump states, valve states)
2. All network features (packet_count, rolling packet rates, packet deltas, network z-scores)
3. Derived features (flow/tank gradients, actuator state changes, rolling std for noise, rolling means for smoothing)

Now that our aligned dataframe has raw physical sensor values, raw network packet counts and labels, we can engineer features to capture behavior over time. We need to consider that ICS processes may change slowly, so we need derivatives and rolling windows.

In [ ]:
# Create a binary anomaly label
full_aligned_dataset['is_anomaly'] = (full_aligned_dataset['label'] != 'normal').astype(int)

In [ ]:
# Encode attack types
full_aligned_dataset['attack_type'] = full_aligned_dataset['label'].where(
    full_aligned_dataset['label'] != 'normal', 'none'
)

In [ ]:
# Train models on normal data only
train = full_aligned_dataset[full_aligned_dataset['label'] == 'normal']
test  = full_aligned_dataset[full_aligned_dataset['label'] != 'normal']

In [ ]:
gc.collect()

365

We need to see packet bursts during attacks, physical disturbances and timing relationships. For building a classifier, we need to drop non-numeric columns, fill NaNs, scale features and conduct a train/test split.

**Derived features**

In [ ]:
# Tank levels, packet_count, actuator deltas get rolling features
rolling_cols = [
    'tank_1','tank_2','tank_3','tank_4','tank_5','tank_6','tank_7','tank_8',
    'packet_count'
]

In [ ]:
# Sort by scenario + time
full_aligned_dataset = full_aligned_dataset.sort_values(
    ['scenario', full_aligned_dataset.index.name]
)

In [ ]:
# Rolling mean per scenario
window = 5  # 5-second rolling window

full_aligned_dataset[[f"{c}_roll_mean" for c in rolling_cols]] = (
    full_aligned_dataset.groupby('scenario')[rolling_cols]
        .rolling(window=window, min_periods=1)
        .mean()
        .reset_index(level=0, drop=True)
)

In [ ]:
# Rolling std per scenario
full_aligned_dataset[[f"{c}_roll_std" for c in rolling_cols]] = (
    full_aligned_dataset.groupby('scenario')[rolling_cols]
        .rolling(window=window, min_periods=1)
        .std()
        .reset_index(level=0, drop=True)
)

In [ ]:
# Rolling deltas
for c in rolling_cols:
    full_aligned_dataset[f"{c}_delta"] = (
        full_aligned_dataset.groupby('scenario')[c].diff()
    )

In [ ]:
# Rolling z-scores
for c in rolling_cols:
    full_aligned_dataset[f"{c}_z"] = (
        (full_aligned_dataset[c] - full_aligned_dataset[f"{c}_roll_mean"]) /
        (full_aligned_dataset[f"{c}_roll_std"] + 1e-6)
    )

Now are data is fully-aligned, cleaned, merged across all scenarios and has been enriched with rolling features to capture both short-term and long-term trends in our CIS. It can be used for anomaly detection.

### 4.4 - Handling Outliers

Models:
1. Regression - remove/winsorize
2. KNN - scale + remove
3. Naive Bayes - extreme values removed
4. SVM - scale + remove
5. DecisionTree - keep
6. Autoencoders = keep
7. GNN - imputation + scale + remove

In [ ]:
# Identify outliers via z-scores
outlier_mask = full_aligned_dataset[[c + "_z" for c in rolling_cols]].abs() > 5

This flags values more than 5 standard deviations above normal behavior.

In [ ]:
# Cap outliers instead of removing rows (winsorization)
for c in rolling_cols:
    full_aligned_dataset[c] = full_aligned_dataset[c].clip(
        lower=full_aligned_dataset[c].quantile(0.001),
        upper=full_aligned_dataset[c].quantile(0.999)
    )

In [ ]:
# Use a robust estimate as the absolute tank max
tank_cols = [f"tank_{i}" for i in range(1, 9)]
tank_max_est = full_aligned_dataset[tank_cols].quantile(0.999)
print(tank_max_est)

tank_1    2014.000000
tank_2    1978.019208
tank_3    3426.000000
tank_4    1196.067228
tank_5     838.000000
tank_6     758.076832
tank_7     812.000000
tank_8     505.000000
Name: 0.999, dtype: float64


In [ ]:
# Fix impossible physical values
for t in tank_cols:
    full_aligned_dataset[t] = full_aligned_dataset[t].clip(
        lower=0,
        upper=tank_max_est[t]
    )

In [ ]:
# Fix actuator glitches (they should not toggle 10 times per second)
bool_cols = full_aligned_dataset.select_dtypes(bool).columns

for col in bool_cols:
    full_aligned_dataset[col] = (
        full_aligned_dataset.groupby("scenario")[col]
        .rolling(window=3, min_periods=1)
        .median()
        .reset_index(level=0, drop=True)
        .astype(bool)
    )

In [ ]:
# Fix packet_count spikes
full_aligned_dataset['packet_count'] = full_aligned_dataset['packet_count'].clip(
    upper=full_aligned_dataset['packet_count'].quantile(0.999)
)

In [ ]:
# Normalize tank values
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
full_aligned_dataset[tank_cols] = scaler.fit_transform(full_aligned_dataset[tank_cols])

Our tank values are clean and normalized, but we do have outliers only in network features.

In [239]:
# Network features to normalize
network_cols = [
    'packet_count', 'packet_delta', 'packet_count_delta',
    'packet_rate_5s', 'packet_rate_30s', 'packet_std_30s',
    'pkt_mean_5s', 'pkt_mean_30s', 'pkt_std_30s',
    'pkt_rate_change', 'packet_count_roll_mean',
    'packet_count_roll_std', 'packet_count_z'
]

flow_cols = ['flow_sensor_1', 'flow_sensor_2', 'flow_sensor_3', 'flow_sensor_4']

In [240]:
# Clip outliers using 0.1% - 99.9% percentiles
for col in network_cols + flow_cols:
    lower = full_aligned_dataset[col].quantile(0.001)
    upper = full_aligned_dataset[col].quantile(0.999)
    full_aligned_dataset[col] = full_aligned_dataset[col].clip(lower, upper)

This removes packet bursts, flow sensor spikes, delta explosions and z-score anomalies without removing rows.

## 5. Data Exploration

### 5.1 - Final Dataset

In [242]:
import matplotlib.pyplot as plt
import seaborn as sns

In [243]:
plt.style.use("seaborn-v0_8")

In [244]:
# Basic dataset summary
print("Shape:", full_aligned_dataset.shape)
print("\nColumns:\n", full_aligned_dataset.columns.tolist())
print("\nInfo:")
print(full_aligned_dataset.info())
print("\nDescribe (numerical):")
print(full_aligned_dataset.describe())

Shape: (10903, 115)

Columns:
 ['tank_1', 'tank_2', 'tank_3', 'tank_4', 'tank_5', 'tank_6', 'tank_7', 'tank_8', 'pump_1', 'pump_2', 'pump_3', 'pump_4', 'pump_5', 'pump_6', 'flow_sensor_1', 'flow_sensor_2', 'flow_sensor_3', 'flow_sensor_4', 'valv_1', 'valv_2', 'valv_3', 'valv_4', 'valv_5', 'valv_6', 'valv_7', 'valv_8', 'valv_9', 'valv_10', 'valv_11', 'valv_12', 'valv_13', 'valv_14', 'valv_15', 'valv_16', 'valv_17', 'valv_18', 'valv_19', 'valv_20', 'valv_21', 'valv_22', 'label_n', 'label', 'packet_count', 'label_change', 'scenario', 'lable_n', 'pump_1_change', 'flow_zscore', 'pkt_zscore', 'packet_rate_5s', 'packet_rate_30s', 'packet_std_30s', 'packet_delta', 'pkt_mean_5s', 'pkt_mean_30s', 'pkt_std_30s', 'pkt_rate_change', 'tank_1_diff', 'flow_1_mean_10s', 'pump_1_changes', 'is_anomaly', 'attack_type', 'tank_1_roll_mean', 'tank_2_roll_mean', 'tank_3_roll_mean', 'tank_4_roll_mean', 'tank_5_roll_mean', 'tank_6_roll_mean', 'tank_7_roll_mean', 'tank_8_roll_mean', 'packet_count_roll_mean', 'ta

Normal traffic dominates (~79%). MITM attacks are the most common attack type, while Scan events are the rarest (14s). Physical faults and DoS attacks are moderately present. Our dataset is normalized and has 98 engineered features.

**Features**
| Raw Physical Features | Network Features     | Rolling Features   | Labels      | Scenario Metadata |
|-----------------------|----------------------|--------------------|-------------|-------------------|
| 8 tanks               | packet_count         | tank rolling means | label       | scenario          |
| 6 pumps               | packet deltas        | tank rolling std   | label_n     | label_change      |
| 4 flow sensors        | packet rolling means | tank deltas        | attack_type |                   |
| 22 valves             | packet rolling std   | tank z-scores      | is_anomaly  |                   |
|                       | packet z-scores      |                    |             |                   |

In [ ]:
# Label distribution
label_counts = full_aligned_dataset['label'].value_counts()

# Bar chart
plt.figure(figsize=(8,4))
sns.barplot(x=label_counts.index, y=label_counts.values, palette="viridis")
plt.title("Label Distribution")
plt.ylabel("Count")
plt.xticks(rotation=45)
plt.show()

# Pie chart
plt.figure(figsize=(6,6))
plt.pie(label_counts.values, labels=label_counts.index, autopct='%1.1f%%')
plt.title("Label Distribution (Pie Chart)")
plt.show()

Normal run is the longest (~57 min.). Attack runs vary in duration.

In [ ]:
# Scenario distribution
scenario_counts = full_aligned_dataset['scenario'].value_counts()

plt.figure(figsize=(8,4))
sns.barplot(x=scenario_counts.index, y=scenario_counts.values, palette="magma")
plt.title("Scenario Distribution")
plt.ylabel("Count")
plt.xticks(rotation=45)
plt.show()

In [ ]:
# Numerical feature histograms
numerical_cols = full_aligned_dataset.select_dtypes(include=['int64', 'float64']).columns

full_aligned_dataset[numerical_cols].hist(figsize=(16,12), bins=30)
plt.suptitle("Numerical Feature Distributions", fontsize=16)
plt.show()

In [ ]:
# Boolean actuator state frequencies
bool_cols = [col for col in full_aligned_dataset.columns if full_aligned_dataset[col].dtype == 'bool']

bool_counts = full_aligned_dataset[bool_cols].sum().sort_values(ascending=False)

plt.figure(figsize=(10,5))
sns.barplot(x=bool_counts.index, y=bool_counts.values, palette="coolwarm")
plt.title("Actuator ON Frequency")
plt.ylabel("Seconds ON")
plt.xticks(rotation=90)
plt.show()

In [ ]:
# Packet count distribution
plt.figure(figsize=(8,4))
sns.histplot(full_aligned_dataset['packet_count'], bins=50, kde=True)
plt.title("Packet Count Distribution")
plt.xlabel("Packets per Second")
plt.show()

In [ ]:
# Correlation heatmap
plt.figure(figsize=(14,10))
corr = full_aligned_dataset.corr(numeric_only=True)
sns.heatmap(corr, cmap="coolwarm", center=0)
plt.title("Correlation Heatmap")
plt.show()

In [ ]:
# Feature behavior by label
plt.figure(figsize=(10,5))
sns.boxplot(data=full_aligned_dataset, x='label', y='packet_count')
plt.title("Packet Count by Label")
plt.xticks(rotation=45)
plt.show()

In [ ]:
# Tank + packet count timeseries
plt.figure(figsize=(14,5))
plt.plot(full_aligned_dataset.index, full_aligned_dataset['tank_1'], label='tank_1')
plt.plot(full_aligned_dataset.index, full_aligned_dataset['packet_count'], label='packet_count', alpha=0.6)
plt.legend()
plt.title("Tank 1 and Packet Count Over Time")
plt.show()

[Provide a brief summary of the final dataset after cleaning and preprocessing. Highlight the most important features and why they are useful.]

[Explore your dataset, summarize and visualize with useful grpahs, like a bar chart, pie chart, etc.) wherever it makes sense.]
[Explore your dataset well along the distribution of data.]
[Numerical and categorical data attributes listed and summarized well.]

number of rows, number of features after cleaning, key important features (flows, pump states, etc.), why each is useful for anomaly detection

histograms of flow rates, correlation maps, timeseries visualizations (before and after attacks), distribution of attack vs. normal

In [ ]:
# Encode categorical labels
aligned['label_n'] = aligned['label'].map({'normal':0, 'MITM':1, 'physical fault':2})

In [ ]:
# Time-series split into train/test
train = aligned[:'2021-04-09 18:50:00']
test = aligned['2021-04-09 18:50:00':]

In [ ]:
# Separate features and labels
X = aligned.drop(columns=['label', 'label_n'])
y = aligned['label_n']

## 6. Regression

[Apply linear and polynomial regression methods for prediction as well as logistical regression for classification decisions.]
[Show every method clearly, apply evaluation metrics to show the correctness of the method, and interpret results well and clearly.]

Regression is not core to anomaly detection.

Possibilities:
1. Linear regression predicting pressure or flow from actuator states
2. Polynomial regression modeling one physical variable
3. Logistic regression for binary classification of anomaly vs. normal

Include R^2, RMSE and a confusion matrix for logistic regression

## 7. Supervised Learning

[Show the application of one or more supervised learning methods (KNN, Naive Bayes, SVM.]
[Apply evaluation metrics and show the correctness of the method(s).]

Use:
1. KNN classifier
2. Naive Bayes
3. SVM

Apply to attack vs. normal OR multiclass attack classification
> Include accuracy, F1 and precision/recall

## 8. Unsupervised Learning

[Show the application of one or more unsupervised learning methods (k-means, hierarchicial clustering, etc.). Apply evaluation metrics and show the correctness of the method(s).]

GNN autoencoder, also include: K-means clustering on sensor features, hierarchical clustering dendogram and silhouette score

## 9. Decision Trees

[Show the application of one or more decision tree methods for classification and/or regression.]
[Apply evaluation metrics and show the correctness of the method(s).]

Use:
1. regular decision tree classifier
2. randomforest
3. xgboost (core to pipeline)

> Include feature importance plots, confusion matrix and ROC curves

## 10. Conclusion

[Lessons learned and future work.]

[Results/Findings - insights, observations, preliminary results from analysis. Outline next steps and give credits and acknowledgements to people.]
What worked best, how GNN AE detected anomalies, how SHAP/GNNExplainer improved interpretability, how XGboost mapped anoamlies to recommended actions, limitations, and future work (reatl-time pipeline, more attacks, acdversarial robustness)

## A. Appendix

### A1 - Previous Notebooks

**[1. Notebook title.]**
<br>
Summary: [what it covers]
<br>
Key Tasks:
[Data analysis, model building, etc.]

Tools/Techniques Used:
[Pandas, NumPy, Scikit-learn, etc.]

Challenges Faced:
[Highlight any challenges you faced while working with the notebooks (e.g., data quality issues, performance issues).]

## A2 - Compilation of Notebooks

[Describe the steps taken to combine all previous notebooks into one final notebook. Mention any adjustments you had to make to ensure the workflow was coherent across all notebooks. Reference any key functions, methods, or classes that were reused.]

AttackOnlyXGBoost.ipynb

BinaryBoost.ipynb

ClassifierXGBoost.ipynb

NetworkDataset_exploration.ipynb

OGdataXGboost.ipynb

OSSIM_Exploration.ipynb

scada_agg.ipynb

EDANetwork1.ipynb

EDAPhysical1.ipynb

EDA_phys_agg_10s.ipynb

GNNae.ipynb

InitialAnomalyDetection.ipynb

NetworkAggregated.ipynb

Phy_att_1.ipynb

Phy_att_3.ipynb

Phy_att_4.ipynb

Phys_att_2.ipynb

ProgRepRegression.ipynb

evaluate.ipynb

exploration.ipynb

learning.ipynb

Physical_data_analysis.ipynb

XGBoost_with_SHAP.ipynb

xgb_model.json

xgboost.ipynb

Explain how you combined everything into one final structured workflow.

Write:

How the notebooks were merged into a single pipeline

The order: ETL → EDA → GNN AE → XAI → XGBoost → Evaluation

Removed redundant code

Unified feature engineering methods

Standardized train/test split

Combined evaluation metrics across models

Built helper functions to avoid repetition

Ensured consistent use of cleaned dataset

Final notebook produces:
1. Clean dataset
2. Anomaly scores
3. Explanations
4. XGBoost recommendations

Notebook: GNNae.ipynb
Summary: Implements the graph neural network autoencoder for anomaly detection using physical + network graph features. Constructs the graph, trains the autoencoder, and evaluates reconstruction error.
Key Tasks:

Graph construction with NetworkX/Neo4j export

GNN autoencoder training

Anomaly score computation

Parameter tuning
Tools: PyTorch Geometric, NumPy, pandas
Challenges: Graph too large → memory limits, had to prune edges and batch nodes.